docs/proposals/client-package-structure.md

```diff
@@ -0,0 +1,295 @@
+a chained function call.
+
+Ideally, no semantic behavior is built into RESTClient, and RESTClient will use
+the Codec it was constructed with for all semantic operations, including turning
+options objects into URL query parameters. Unfortunately, that is not true of
+today's RESTClient, which may have some semantic information built in. We will
+remove this.
+
+RESTClient should not make assumptions about the format of data produced or
+consumed by the Codec. Currently, it is JSON, but we want to support binary
+protocols in the future.
+
+The Codec would look something like this:
+
+```go
+type Codec interface {
+  Encode(runtime.Object) ([]byte, error)
+  Decode([]byte]) (runtime.Object, error)
+
+  // Used to version-control query parameters
+  EncodeParameters(optionsObject runtime.Object) (url.Values, error)
+
+  // Needed only on the server, not by RESTClient
+  DecodeParameters(query url.Values, destOptionsObject runtime.Object) error
+}
+```
+
+There should be one codec per version. RESTClient is *not* responsible for
+converting between versions; if a client wishes, they can supply a Codec that
+does that. But RESTClient will make the assumption that it's talking to a single
+group/version, and will not contain any conversion logic. (This is a slight
+change from the current state.)
+
+As with Transport, it is expected that 3rd party providers following the api
+conventions should be able to use RESTClient, and will not need to implement
+their own.
+
+Action items:
+* Split out of the current client package. (@krousey)
+* Possibly, convert to an interface (currently, it's a struct). This will allow
+  extending the error-checking monad that's currently in request.go up an
+  additional layer.
+* Switch from ParamX("x") functions to using types representing the collection
+  of parameters and the Codec for query parameter serialization.
+* Any other Kubernetes group specific behavior should also be removed from
+  RESTClient.
+
+### Mux layer
+
+The client muxer layer has a map of group/version to cached RESTClient, and
+knows how to construct a new RESTClient in case of a cache miss (using the
+discovery client mentioned below). The ClientMux may need to deal with multiple
+transports pointing at differing destinations (e.g. OpenShift or other 3rd part
+provider API may be at a different location).
+
+Tentative name: ClientMux
+
+Action items:
+* Move client cache out of kubectl libraries into a more general home.
+* TODO: a mux layer may not be necessary, depending on what needs to be cached.
+  If transports are cached already, and RESTClients are extremely light-weight,
+  there may not need to be much code at all in this layer.
+
+### High-level: Individual typed
+
+Our current high-level client allows you to write things like
+`c.Pods("namespace").Create(p)`; we will insert a level for the group.
+
+That is, the system will be:
+
+`clientset.GroupName().NamespaceSpecifier().Action()`
```

Comment:
@caesarxuchao It seems the resource also takes a namespace in his example on line 148. So it would be like `clientset.GroupName().Pods("namespace").Create(p)` if I'm understanding correctly.